### PR TITLE
A few last minute Java-side changes + documentation

### LINKED ISSUE #16439: Provide API for writing custom marshallers for Java routing
Currently, you need to derive from `impl.MarshallerImpl` and `impl.UnmarshallerImpl`.

### PR DESCRIPTION
**Pull request by jrudolph**
Fixes #16439.

/cc @sirthias

### CODE REVIEW

File: akka-docs-dev/rst/java/http/routing-dsl/marshalling.rst [hunk 1]
```diff
@@ -3,4 +3,41 @@
...
+-----------
+
+On the server-side marshalling is used to convert a application-domain object to a response (entity). Requests can
+contain an ``Accept`` header that lists acceptable content types for the client. A marshaller contains the logic to
+negotiate the result content types based on the ``Accept`` and the ``AcceptCharset`` headers.
+
+Marshallers can be specified when completing a request with ``RequestContext.completeAs`` or by using the ``BasicDirectives.completeAs``
+directives.
+
+These marshallers are provided by akka-http:
```
Comment (ktoso) [inline akka-docs-dev/rst/java/http/routing-dsl/marshalling.rst]: Minor:  We should consolidate up how we write akka-http or Akka Http or Akka HTTP at some point.

Comment (jrudolph) [inline akka-docs-dev/rst/java/http/routing-dsl/marshalling.rst]: Yep, good point.

File: akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala [hunk 2]
```diff
@@ -15,46 +15,149 @@ import scala.collection.immutable
...
   @varargs
   def redirectToTrailingSlashIfMissing(redirectionStatusCode: StatusCode, innerRoute: Route, moreInnerRoutes: Route*): Route =
     RedirectToTrailingSlashIfMissing(redirectionStatusCode)(innerRoute, moreInnerRoutes.toList)
+
+  /**
+   * If the request path ends with a slash, redirect to the same uri without trailing slash in the path.
+   *
+   * '''Caveat''': [[pathSingleSlash]] directive will not match inside of this directive.
+   */
   @varargs
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala]: all these docs are from scaladsl, right? Thanks for copying over.

File: akka-http/src/main/scala/akka/http/javadsl/server/directives/RouteDirectives.scala [hunk 3]
```diff
@@ -1,18 +0,0 @@
...
-import akka.http.javadsl.model.{ StatusCode, Uri }
-
-abstract class RouteDirectives extends RangeDirectives {
-  /**
-   * Completes the request with redirection response of the given type to the given URI.
-   *
-   * The ``redirectionType`` must be a StatusCode for which ``isRedirection`` returns true.
-   */
-  def redirect(uri: Uri, redirectionType: StatusCode): Route = Redirect(uri, redirectionType)
-}
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/javadsl/server/directives/RouteDirectives.scala]: +1

File: akka-http/src/main/scala/akka/http/javadsl/server/directives/BasicDirectives.scala [hunk 4]
```diff
@@ -64,6 +64,13 @@ abstract class BasicDirectives {
     }
 
   /**
+   * Completes the request with redirection response of the given type to the given URI.
+   *
+   * The ``redirectionType`` must be a StatusCode for which ``isRedirection`` returns true.
+   */
+  def redirect(uri: Uri, redirectionType: StatusCode): Route = Redirect(uri, redirectionType)
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/javadsl/server/directives/BasicDirectives.scala]: +1 here is nicer

File: akka-docs-dev/rst/java/http/routing-dsl/directives/index.rst [hunk 5]
```diff
@@ -3,4 +3,57 @@
...
+
+These classes of directives are currently defined:
+
+BasicDirectives
+  Contains methods to create routes that complete with a static values or allow specifying :ref:`handlers-java` to
+  process a request.
+
+CacheConditionDirectives
+  Contains a single directive ``conditional`` that wraps its inner route with support for Conditional Requests as defined
+  by http://tools.ietf.org/html/draft-ietf-httpbis-p4-conditional-26.
```
Comment (ktoso) [inline akka-docs-dev/rst/java/http/routing-dsl/directives/index.rst]: is this a clickable link in the resulting doc?

**Comment by ktoso** [comment]
LGTM, minor question about if the link in directives/index is properly clickable?

**Comment by 2m** [comment]
LGTM

File: akka-docs-dev/rst/java/http/routing-dsl/directives/index.rst (hunk 5 above)
Comment (jrudolph) [inline akka-docs-dev/rst/java/http/routing-dsl/directives/index.rst]: Yeah, good point, it is clickable but an outdated link...

Comment (jrudolph) [inline akka-docs-dev/rst/java/http/routing-dsl/directives/index.rst]: Updating..

File: akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala (hunk 2 above)
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala]: Yes, mostly copied over.

**Comment by jrudolph** [comment]
Got a new link.

**Comment by jrudolph** [comment]
I added a few final PathMatcher / PathDirective Javadoc and documentation details. That's all I have for today ;). I mocked out example bits that would conflict with #17988 so that we don't need another roundtrip. I'll fix it tomorrow.

**Comment by ktoso** [comment]
LGTM, thanks

**Pull request merged**